hummingbot/connector/derivative/bybit_perpetual/bybit_perpetual_api_order_book_data_source.py

```diff
@@ -332,148 +80,264 @@ async def listen_for_subscriptions(self):
-
-                if event_type == "delta":
-                    diff_msg: OrderBookMessage = BybitPerpetualOrderBook.diff_message_from_exchange(
-                        msg=order_book_message,
-                        timestamp=int(order_book_message["timestamp_e6"]) * 1e-6,
-                        metadata={"trading_pair": trading_pair})
-                    output.put_nowait(diff_msg)
-            except asyncio.CancelledError:
-                raise
-            except Exception as ex:
-                self.logger().error(f"Unexpected error ({ex})", exc_info=True)
+    async def _get_connected_websocket_assistant(self, ws_url: str) -> WSAssistant:
+        ws: WSAssistant = await self._api_factory.get_ws_assistant()
+        await ws.connect(
+            ws_url=ws_url, message_timeout=CONSTANTS.SECONDS_TO_WAIT_TO_RECEIVE_MESSAGE
+        )
+        return ws
 
-    async def listen_for_order_book_snapshots(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
-        """
-        Periodically polls for orderbook snapshots using the REST API.
-        """
-        await self._sleep(self._ORDER_BOOK_SNAPSHOT_DELAY)
+    async def _subscribe_to_channels(self, ws: WSAssistant, trading_pairs: List[str]):
+        try:
+            symbols = [
+                await self._connector.exchange_symbol_associated_to_pair(trading_pair=trading_pair)
+                for trading_pair in trading_pairs
+            ]
+            symbols_str = "|".join(symbols)
+
+            payload = {
+                "op": "subscribe",
+                "args": [f"{CONSTANTS.WS_TRADES_TOPIC}.{symbols_str}"],
+            }
+            subscribe_trade_request = WSJSONRequest(payload=payload)
+
+            payload = {
+                "op": "subscribe",
+                "args": [f"{CONSTANTS.WS_ORDER_BOOK_EVENTS_TOPIC}.{symbols_str}"],
+            }
+            subscribe_orderbook_request = WSJSONRequest(payload=payload)
+
+            payload = {
+                "op": "subscribe",
+                "args": [f"{CONSTANTS.WS_INSTRUMENTS_INFO_TOPIC}.{symbols_str}"],
+            }
+            subscribe_instruments_request = WSJSONRequest(payload=payload)
+
+            await ws.send(subscribe_trade_request)  # not rate-limited
+            await ws.send(subscribe_orderbook_request)  # not rate-limited
+            await ws.send(subscribe_instruments_request)  # not rate-limited
+            self.logger().info("Subscribed to public order book, trade and funding info channels...")
+        except asyncio.CancelledError:
+            raise
+        except Exception:
+            self.logger().exception("Unexpected error occurred subscribing to order book trading and delta streams...")
+            raise
 
+    async def _process_websocket_messages(self, websocket_assistant: WSAssistant):
+        channels = [
+            self._diff_messages_queue_key, self._trade_messages_queue_key, self._funding_info_messages_queue_key
+        ]
```

Comment:
This could be a list in `PerpetualAPIOrderBookDataSource`.

Reply:
Addressed